### DISCUSSION TITLE
Evaluate RabbitMQ Microservices Adapter Pattern Architecture

### DISCUSSION THREAD

**Discussion by WonderPanda** [golevelup/nestjs#235] [Ideas]
One of the goals of this library should be to support as many of the built in NestJS enhancers as possible when building apps that use RabbitMQ. Currently `Interceptors` work, but it would be ideal if people were also able to use `Guards` or `Pipes`.

We should integrate properly with the NestJS external context creation process to leverage the full enhancer pipeline.

Additionally, there might be some really big wins for us to consider turning this into an actual instance of  [Custom Transport](https://docs.nestjs.com/microservices/custom-transport) Microservices implementation. Turning this into a microservice could also help provide some better design patterns for managing the underlying AMQP connection. This could prove useful for the ideas proposed in #233

**Comment by underfisk** [comment]
This would be really great, most of the microservices want a transport layer and use something similar to the official microservices API provided by Nestjs, i really do agree with this
Either the package turns into using microservices adapter and support the current services layer or provides a major with only microservices adapter, i do think it depends because there may be people using this package

**Reply by WonderPanda** [reply]
I think we should consider approaching a redesign using a breaking major change. I've officially started thinking about v2.x roadmap in #227. If we're going to make big changes I'd like to try to lump them together so that 2.x can be a stable release and we don't have to do a whole bunch more breaking changes in short succession.

If there's an opportunity that there are some short term wins that can be incorporated into the existing package without introducing any breaking changes we could definitely consider it. Sometimes it ends up taking more work trying to make sure that all potential breaking changes are avoided

**Reply by underfisk** [reply]
I agree
